File: backend/backtesting/engine.py
from datetime import datetime
from collections import deque

from backtesting.portfolio import Portfolio

class BacktestResult:
    def __init__(self):
        self.history = []
        self.trades = []
        self.final_value = 0
        self.profit_loss = 0
        self.total_return = 0
        self.num_trades = 0
        self.annualized_return = 0
        self.max_drawdown = 0
        self.win_rate = 0

class BacktestEngine:
    def __init__(self, strategy, startDate, endDate):
        self.strategy = strategy
        self.startDate = self._to_datetime(startDate)
        self.endDate = self._to_datetime(endDate)

    def _to_datetime(self, d):
        if isinstance(d, datetime):
            return d

        return datetime.strptime(d, "%Y-%m-%d")
    
    def run(self, data, initial_cash):
        portfolio = Portfolio(initial_cash)

        result = BacktestResult()

        peak = None
        max_drawdown = None
        drawdown = None

        for i, (date, row) in enumerate(data.iterrows()):
            signal, qty = self.strategy.decide(i, row, len(data))

            price = row["Close"]

            if signal == "BUY":
                bought = portfolio.buy(price, qty)
                if bought > 0:
                    result.num_trades += 1
                    result.trades.append({
                        "Date": str(date),
                        "Portfolio Value": portfolio.value(price),
                        "Cash": portfolio.cash,
                        "Shares": portfolio.shares,
                        "Action": "Buy",
                        "Price": float(price),
                        "Quantity": int(bought)
                    })
            elif signal == "SELL":
                sold = portfolio.sell(price, qty)
                if sold > 0:
                    result.num_trades += 1
                    result.trades.append({
                        "Date": str(date),
                        "Portfolio Value": portfolio.value(price),
                        "Cash": portfolio.cash,
                        "Shares": portfolio.shares,
                        "Action": "Sell",
                        "Price": float(price),
                        "Quantity": int(sold)
                    })

            result.history.append({
                "Date": str(date),
                "Portfolio Value": portfolio.value(price),
                "Cash": portfolio.cash,
                "Shares": portfolio.shares
            })

            if peak == None or portfolio.value(price) > peak:
                peak = portfolio.value(price)

            if peak > 0:
                drawdown = (peak - portfolio.value(price))/peak
            else:
                drawdown = 0

            if max_drawdown == None or drawdown > max_drawdown:
                max_drawdown = drawdown
        
        years = (self.endDate - self.startDate).days / 365.25

        result.final_value = portfolio.value(price)
        result.profit_loss = result.final_value - initial_cash
        result.total_return = result.profit_loss/initial_cash
        result.annualized_return = ((result.final_value/initial_cash) ** (1/years)) - 1
        result.max_drawdown = max_drawdown
        result.win_rate = portfolio.win_rate()

        return result

File: backend/backtesting/engine.cpp
#include <iostream> 
#include <chrono>
#include <vector>
#include <string>
#include <cmath>
#include "engine.h"
#include "portfolio.h"

Result Engine::run(const std::vector<MarketData>& data, double initial_cash){
        Portfolio portfolio(initial_cash);
        Result result = Result();

        double peak = 0;
        double max_drawdown = 0;
        double drawdown = 0;
        double price = 0;

        for(int i = 0; i < data.size(); i++){
            const MarketData& row = data[i];
            Decision dec = strategy.decide(i, row, data.size());
            price = row.close;

            if(dec.signal == "BUY"){
                int bought = portfolio.buy(price, dec.qty);
                if(bought > 0){
                    Trades trade = {
                        row.date,
                        portfolio.value(price),
                        portfolio.cash,
                        portfolio.shares,
                        "Buy",
                        double(price),
                        int(bought)
                    };
                    result.trades.push_back(trade);
                    result.num_trades ++;
                }
            }
            else if(dec.signal == "SELL"){
                int sold = portfolio.sell(price, dec.qty);

                if(sold > 0){
                    Trades trade = {
                        row.date,
                        portfolio.value(price),
                        portfolio.cash,
                        portfolio.shares,
                        "Sell",
                        double(price),
                        int(sold)
                    };
                    result.trades.push_back(trade);
                    result.num_trades++;
                }
            }

            History hist = {
                row.date,
                portfolio.value(price),
                portfolio.cash,
                portfolio.shares,
            };
            result.history.push_back(hist);

            if(portfolio.value(price) > peak){
                peak = portfolio.value(price);
            }

            if(peak > 0){
                drawdown = (peak - portfolio.value(price))/peak;
            }
            else{
                drawdown = 0;
            }

            if(drawdown > max_drawdown){
                max_drawdown = drawdown;
            }
        }
        double years = std::chrono::duration<double>(endDate - startDate).count() / (365.25 * 24 * 60 * 60);
        
        result.final_value = portfolio.value(price);
        result.profit_loss = result.final_value - initial_cash;
        result.total_return = result.profit_loss/initial_cash;
        result.annualized_return = std::pow(result.final_value / initial_cash, 1.0 / years) - 1;
        result.max_drawdown = max_drawdown;
        result.win_rate = portfolio.winRate();

        return result;
}